### ISSUE TITLE
Please consider adding .direct() or equivalent to getCause() to complement 

### ISSUE THREAD

**Issue by theboomer** [SpongePowered/SpongeAPI#931]
For things like the BlockChangeEvent.Place, it gets triggered when ground is hoed or sheep eat grass, with a root cause of player or sheep  (or player again, if a player spawned the sheep)...  But then if the 'placed' farmland dries out to dirt, the root cause is a player, and the other cause is a spongeblocksnapshot; same for the sheep-eaten dirt returning to grass, it is root cause of player, other cause of spongeblocksnapshot,  { and nowhere a sheep in the cause list???  A ponderance for another day...}

The root cause lets us access right away the object that is the root cause, but if the action is one of a chain, the 'direct action' (aka the spongeblocksnapshot dirt-->grass // farmland-->Dirt transition ) exists without a .. direct .. way to obtain it clearly.

For example, I want to exit the block-place event if it was not a direct/inferred direct(farmland) placement by a player object, otherwise, i will do things with the player object that directly placed a block:
   getCause().root returns player object but it also returns it when grass grows on blocks his sheep ate, or grass grows on blocks that were under blocks he broke moments earlier, or when grass converts to dirt for blocks he covered with a solid block moments earlier., or when the piece of farmland he hoed dries out to dirt; or when that dried out dirt further converts to grass - not the same thing as 'directly placed the block of grass' when it happens 3 mins later and 50 blocks away from the player...
  getCause().first(Player.class).isPresent() returns true, as it is the player  BUT
  getCause().last(Player.class).isPresent() also returns true, as he is both the first and last player object in the cause system.
  obviously .any(Player.class) and .contains(Player.class) will reflect the undesired root case..
  It gets overly complicated with having to test for the object that comes after or before a particular class - testing for classes after the player would return things other than player as root, as that blows away the one-and-only cause-case of a player directly placing a block...
  
It seems that the the shortest/easiest way at the 'direct' cause is the getCause().all().get(0);  and that root() is the last item in the all() list returned.  So would it please be possible to add a friendly method called 'direct()' or something similar that returns the get(0) object, regardless of what it is, in the same way that root() returns the object from the other end, to make for left/right - up/down - first/last before/after complementation  .  I kept having my event triggered long after the fact in one of my plugins because i was detecting 'a player object' when grass grew or died minutes later, because i had used .get(0) elsewhere earlier, but was not consistant with it in my coding of other plugins, as its not intuitive as seeing something like .direct()

Basically, ".root()" is a special case of an element in the list of causes with a special identity, being the last element; but so is the other end of the list also a special case element, and deserves an equivalent friendly, readable identity like direct()  or some other term you may find more appropriate.
.agent(), .provocateur(), .blame(), .accuse(), .suspect(), .patsy(), .fallguy(), .redhanded(), .triggerer() etc - if "direct()" is too 'inaccurate' of a term for other scenarios...

**Comment by Deamon5550** [comment]
Uhh I think you have it backwards, `root()` already is returning the first or 'direct' cause (see [Cause.java#L332](https://github.com/SpongePowered/SpongeAPI/blob/master/src/main/java/org/spongepowered/api/event/cause/Cause.java#L332))

**Comment by Zidane** [comment]
First, it would help if you could be a bit more...direct in your issues so I can understand what you are trying to ask of us. Not to be mean or anything, its just hard to see where your issue is and where is the ramble?

So let me break this down:

1). First, if a dirt block the Sheep caused due to grass eat is turned back into grass, we cannot know this without storing the sheep UUID across ticks...but what if the sheep is dead? Causes are powerful but there is a line we draw. Its simple enough for you, as a plugin, to handle this yourself.
2). Your root is the FIRST object. Do you want the LAST object instead?

```
        @Override
        public Optional<?> root() {
            return Optional.of(this.cause[0]);
        }
```

**Comment by theboomer** [comment]
Sorry for unclearness - I was trying to make sense of out some conflicting dumps in my logs for particular events, and rally just should say "Run this code, do this step, look at the logs" as they appeared to be inconsistant.

Primarily yes, i incorrectly mixed up the 0th and nth element in my confusion and tiredness, wanting "the opposite end of the chain from the root cause" 

I dont care about the sheep, and i think something got scrambled in my log regarding sheep -- my concern was over the dirt blocks turning to grass, and grass blocks turning to dirt  minutes after covering or uncovering them by player action, and something not reporting/testing right that i now cant see in the log printouts again, so it might have been a mislabled output statement confusing me.

It SEEMS that yes, in these cases, the player would be the root cause indeed, but not the direct cause finality-cause of the block of dirt turning to grass at xyz position at time t minutes after the player has left the area, and is the all() nth case.

So just distill this down to " please add a complementary method to root() that does the same thing at the opposite end of the chain , with one method returning 0th, the other returning nth

**Comment by Zidane** [comment]
@gabizou 

I see no issues in this. Should it still be called direct (since it is the last one in the chain before the action) or last? I'd rather avoid last though.

**Comment by gabizou** [comment]
The thing is, the root cause is the direct cause, anything afterwards should be considered the "indirect" or "aid" cause.

**Comment by Zidane** [comment]
@gabizou then we should name "root" direct and add an in-direct.

Makes a lot more sense to me then.

**Comment by gabizou** [comment]
@Zidane What is indirect returning?

**Comment by Deamon5550** [comment]
`root` at least has some contract around being the most immediate cause of the event. Everything after varies from indirect causes to random pieces of contextual information to the last person to notify the block even if they weren't involved in the event at all. So we can't really put a name like `indirect` on it.

**Comment by Zidane** [comment]
@gabizou 

last element in cause array.

**Comment by Zidane** [comment]
@Deamon5550 

Good point, any other name ideas?

**Comment by Deamon5550** [comment]
@Zidane not having a method at all? I mean we have `all` for a reason, the last element is not special in any way other than it happens to be *last*.

**Comment by theboomer** [comment]
Root is Not the same as Direct though...

Direct cause for replacing window: Rock thrown through it
Root cause: Lousy parenting

Direct cause for sample loss: dissolved in wrong solution
Root cause: Analyst failed to use the correct procedure

Direct cause for extra charges on Visa bill: $250 worth of cleaning supplies
Root cause: Basement sewer drain backed up

Direct cause for ChangeBlockEvent.Place thrown: Dirt block spontaneously changing into Grass due to environmental gamrules
Root cause for ChangeBlockEvent.Place thrown: Player had done something 97 seconds earlier to generate the block of dirt in that location



'''
[15:12:51] [Server thread/INFO] [STDOUT/]: [com.prennet.boomod.sinkholes.Sinkholes:onTrigger:32]: org.spongepowered.api.event.block.ChangeBlockEvent$Place{game=org.spongepowered.mod.SpongeModGame@769c21ff, cancelled=false, cause=Cause{[EntityPlayerMP['BooMod'/361, l='world', x=54.16, y=78.78, z=503.26], SpongeBlockSnapshot{worldUniqueId=d61cff81-2f4f-4f73-8fa2-eada402dd9a0, position=(57, 73, 502), blockState=minecraft:grass[snowy=false], extendedState=minecraft:grass[snowy=false]}]}, transactions=[Transaction{original=SpongeBlockSnapshot{worldUniqueId=d61cff81-2f4f-4f73-8fa2-eada402dd9a0, position=(57, 73, 503), blockState=minecraft:dirt[snowy=false,variant=dirt], extendedState=minecraft:dirt[snowy=false,variant=dirt]}, default=SpongeBlockSnapshot{worldUniqueId=d61cff81-2f4f-4f73-8fa2-eada402dd9a0, position=(57, 73, 503), blockState=minecraft:grass[snowy=false], extendedState=minecraft:grass[snowy=false]}, custom=null, valid=true}], targetWorld=net.minecraft.world.WorldServer@76fd52ab}
[15:12:51] [Server thread/INFO] [STDOUT/]: [com.prennet.boomod.sinkholes.Sinkholes:onTrigger:35]: Event cause EntityPlayerMP['BooMod'/361, l='world', x=54.16, y=78.78, z=503.26]
[15:12:51] [Server thread/INFO] [STDOUT/]: [com.prennet.boomod.sinkholes.Sinkholes:onTrigger:35]: Event cause SpongeBlockSnapshot{worldUniqueId=d61cff81-2f4f-4f73-8fa2-eada402dd9a0, position=(57, 73, 502), blockState=minecraft:grass[snowy=false], extendedState=minecraft:grass[snowy=false]}
[15:12:51] [Server thread/INFO] [STDOUT/]: [com.prennet.boomod.sinkholes.Sinkholes:onTrigger:37]: Root was Optional[EntityPlayerMP['BooMod'/361, l='world', x=54.16, y=78.78, z=503.26]]

**Comment by theboomer** [comment]
Last IS Special  ... it is what caused the event to fire, it is the consequence of the bad parenting, it is the consequence of the sewer backup, it is the consequence of the dirt block formed 97 seconds earlier doing what comes naturally in its environment.]

If this is the ONLY case in the whole game where spontaneous block change events due to grass growth / soil dehydration etc occur  which happen to be the thing that occurs at the very moment that an event is throw, for which the root cause is not that thing that very moment, and for which coincidentaially the last thing in the cause list happens to be the very thing that throws the event trigger, then sure "last is nothing special" it is just coindcidental in this case...

If there is a scenario where more causes would exist in a chain, such that all(0) is the root that happened 48 seconds earlier, all(1) is something that happened 11 seconds earlier, all(2) is something that happened a nanosecond earlier, all(3) is something that happened 17 seconds earlier -- then absolutely, the last one is not the one that happened 'to fire the event' 
Which makes the job even stupider and harder for us to figure out what was the actual actual that triggered the event to respond to, and would even MORE GREATLY demand a special process to identify what triggered the event to fire at the precise moment it did...

But i do not think that is the case by design, i think by design your last object IS the trigger object, and in this example where the root object could be halfway across the world or even in a different dimension or hell, logged out even, at the moment when that dirt spontaneously changes to grass...  how does that compare to root cause = player when player is constructing a box out of stone?  It doesn't.

Thus, Root is not Direct and last is not "no special case"

**Comment by Deamon5550** [comment]
That is not quite the definition of root cause that we're using. By the formal definition the root cause would be some point in the chain where *"an intervention could reasonably be implemented to improve performance or prevent an undesirable outcome"*. Which doesn't really apply since **our cause chain is more contextual than a specific chain of actions** leading up to the current event.

I think perhaps the real issue with what we have currently is in this example of an event thrown from a dirt block changing to grass after a player has uncovered it is that the player is the first object in the cause chain. I think in this case that the chain should be padded with some other object (something like some value to say that this is natural grass growing, like the `Reason`s in the original cause rework) so that the player is not the first cause.

**Comment by gabizou** [comment]
I'm just going to provide some insight with how Causes are structured for use in DamageEvents.

In Damage, the *direct* `DamageSource` is an object, any additional objects within the cause are "associated" with the `DamageSource`.
This is the `Cause` associated when an `Entity` is taking fire damage due to standing in or close enough to a fire block:
```
[spongetest]: ======== DamageEvent START ========
[spongetest]: Cause: Cause[{Name=Source, Object={BlockDamageSource{Name=inFire, Type=fire, BlockSnapshot=SpongeBlockSnapshot{worldUniqueId=f3031705-ef50-4eb2-ae1b-323c71dd3bb1, position=(232, 63, 223), blockState=minecraft:fire[age=1,alt=false,east=false,flip=false,north=false,south=false,upper=2,west=false], extendedState=minecraft:fire[age=1,alt=false,east=false,flip=false,north=false,south=false,upper=2,west=false]}, Location=Location{(232.0, 63.0, 223.0) in net.minecraft.world.WorldServer@7d85e5d5}}}}]
[spongetest]:  ---------------------------------
[spongetest]: Base Damage:1.0
[spongetest]: Modifiers:
[spongetest]: Final Damage: 1.0
[spongetest]: ======= DamageEvent Finish ========

```

As you can see in that, the `Cause` includes a lot of already useful information in a single object, but there's no actual use of the other objects, in fact, the extra objects that would otherwise be included for events like when a player attacks, those objects are just "aides" at that point.

**Comment by theboomer** [comment]
if indeed this is the only condition/example where the root cause is not the actual nanosecond trigger sure - perhaps its just written out backwards for this event?

If there is absolutely otherwise no correlation between the last element in cause and anything meaningful in general, and it just happens in this case to hold meaning because of the backwards output of this event, or because of the "only two objects in the cause chain, one is root, the other is not root" but "not root" is not in this case the intentional same as "last in chain" ...  and is more of a "there are things that are related, but, they could be listed in any random order so the 3rd and last one in the list is no more or less special than the 2nd one in the list" then sure, there is no need/requirement for some function to directly return the last-object in the chain ever.  But then here yes, absolutely, please change this process so that the player is not being reported as root cause for the dirt changing to grass, or grass changing to dirt after he covers it in a block 74 seconds earlier, or the farmland dehdrating to dirt 84 seconds later...

**Comment by JBYoshi** [comment]
([Quote by Deamon5550](https://github.com/SpongePowered/SpongeAPI/issues/931#issuecomment-159100411))
> Everything after [the root cause] varies from indirect causes to random pieces of contextual information to the last person to notify the block even if they weren't involved in the event at all.

This is the one part I don't like about the cause system. Sometimes the root cause is not related to causing the event. Sometimes the "last" cause is a *means* of causing the event (e.g. `ItemStack`), not the cause itself (e.g. `Player`).

Even with a perfect setup, causes cannot be used to determine what truly caused the event - if a player *knowingly* activated some TNT, then they ought to be the "direct" cause, but if they *unknowingly* activated some TNT, then they ought *not* to be the "direct" cause (if a plugin blocked explosions not caused by ops, then someone could still troll an admin by getting them to stand on a pressure plate hooked up to some TNT).

**Comment by Deamon5550** [comment]
@JBYoshi Yeah for sure we need to look at the consistency across the events we're throwing to be sure that the root cause is always the 'direct' action which caused the event, everything else is just context. As for your TNT example, really the protection plugin should be blocking TNT based on the 'owner' tag we attach to blocks and make available as a NamedCause and thus would only explode TNT actually placed down by the admin.

In a perfect world I feel that you should always be able to trust the root cause, but anything else you should take with a grain of salt. Of course with plugins, mods, and the sheer variety of events this can't always be true.

**Comment by theboomer** [comment]
Meaning there should be one more piece of information provided with the event then, that being the actual dang 'thing' that triggered the generation of that event at the precise moment it did - if root cause and cause chains have absolutely no order to them or way to make sense of the things spewn out, they are just .. related things that happened -- then its even more critical that the precise reason the event was thrown should be a special return method in the event as well

Some way to distinguish when the event ChangeBlockEvent.Place happens , there is a clean, clear, simple way to determine that it was due to the player placing the block down at that moment.  Or pull it out ChangeBlockEvent.Environmental  and respond only to block state changes of grass growth\decay, soil dehydration so that Place  root cause of player will always mean "that nanosecond ago, this player slapped down the block"

**Comment by gabizou** [comment]
@theboomer Can you print out what the cause is for the even you're worried about, in this case `ChangeBlockEvent.Place`?

**Comment by theboomer** [comment]
I already did ^^^

**Comment by gabizou** [comment]
Ah, I see it. You should know there's a difference between ' and `. Namely that the accent is used to mark things as code.

**Comment by theboomer** [comment]
Yeah, was surprised it didnt work .. i can edit it

**Comment by gabizou** [comment]
Closing in favor of resolving #948
